### ISSUE TITLE
Cache quota shouldn't be computed until we know we need to write to the cache

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#996]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/996)
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Apr 15, 2019, 11:44

## Summary

[//]: # (Summarize the bug encountered concisely)

Currently cache cas initilization calculates the cache quota. This information is only needed when we will be writing to the cache, and as such should not be calculated until the last possible moment since it is expensive to calculate.

For example. on Windows WSL, it could increase `bst show` time to element load by 10-20 seconds, when the rest of the operations in `bst show` would take roulghy 0.2s.

## Steps to reproduce

[//]: # (How one can reproduce the issue - this is very important)

- Have many things in your cache
- `bst show`, consider the time before the "loading element" message.

## What is the current bug behavior?

[//]: # (What actually happens)

Cas quota is computed unneccessarily for read only operations on the cache.

For example, the following commands don't need to write anything and we could therefore save 20s on their invocation:

- `bst show`
- `bst workspace *`
- `bst shell`
- `bst push`
- `bst source {checkout,push}`
- `bst checkout`
- `bst artifact log`
- `bst artifact delete` (maybe? arguably)

## What is the expected correct behavior?

[//]: # (What you should see instead)

Instantaneous startup, we don't care about the cache size in those scenarios, we shouldn't be losing time computing them.

## Relevant logs and/or screenshots

[//]: # (Paste any relevant logs - please use code blocks ``` to format console output, logs, and code as it's hard to read otherwise. You can also add a Snippet and link it here. Check the markdown giude on Gitlab for further tips)

Timings from the import of "bst" to the moment in code when this was called.

```
cli: function entry 0.19205307960510254
app: context loaded 0.19537138938903809
app: platform instantiated 0.20264530181884766
app: logger ready 0.20321416854858398
base cache: init 0.20326685905456543
base cache: cas cache initialized 0.20374846458435059
base cache: cas quota set 11.165070295333862 <<< CACHE QUOTA COMPUTED
```

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @danielsilverstone-ct]](https://gitlab.com/danielsilverstone-ct) on Apr 15, 2019, 15:18

As a rough outline of what I'd suggest to do:

1. Ensure that nothing `_calculate_cache_quota()`s during startup
2. Add a flag to `CASQuota` which indicates that the quota has not yet been calculated
3. Add a public method to `CASQuota` to set/clear the flag as appropriate and `_calculate_cache_quota()`
4. `assert` in all the methods which could *ADD* to the CAS that the flag indicates the quota has been calculated
5. Determine the latest place possible in all flows which might write to the cache, where we can call that method, and call it.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 16, 2019, 10:31

Not sure about this, have you considered how this factors into the logging aspects ?

What will be printed showing the cache usage at beginning of session before the first write to the artifact cache ?

I mean, in the cases listed above we don't need to recalculate it, I'm all for this, but if we don't even have a cache_size file stored maybe we should still run it and ensure the logging is sane ?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @danielsilverstone-ct]](https://gitlab.com/danielsilverstone-ct) on Apr 16, 2019, 11:05

[[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) Can you confirm if you have a valid cache_size file present in the above tests?

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) Our focus is on not wasting time when it's not really necessary.  Is the cache size information necessary in situations which won't trigger consumptive changes in cache size?  If you really want to give the information to the user then why not have a `bst cache-status` command instead which can do the work in the absence of any other need?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 16, 2019, 11:28

> [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) Our focus is on not wasting time when it's not really necessary. Is the cache size information necessary in situations which won't trigger consumptive changes in cache size? If you really want to give the information to the user then why not have a `bst cache-status` command instead which can do the work in the absence of any other need?

I don't think a separate command is a good idea for this at all (rather it could be nice to add, but not as a replacement for live feedback and logging about cache quota usage).

Right now, we print the cache usage as a percentage value in the status bar during any session, this updates dynamically as the cache usage increases (or decreases during a cleanup) and is very nice to have - also we have a summary of *all relevant information* at the beginning of any session, this should really include as much information as possible for any post build diagnosis, I believe that it is useful to have the current cache usage at the start of a build for instance and that we shouldn't remove that.

Now what I think is this:
* My concern is probably moot because, it doesnt cost anything to read the cached `cache_size` file at startup time, and that file is almost always certainly there unless we are running buildstream for the first time, in which case it is inexpensive to create because the cache will be empty anyways.
* Currently there is a part in the scheduler which blocks any activity from happening until we have forcefully recalculated cache size, this is only important for commands which risk adding anything to the cache, so this can be safely disabled for sessions which do not add anything to the cache

Essentially, I only want to make sure that if we're discarding parts of initialization, we're observing all of the side effects it might have, including logging, before discarding/optionalizing parts of initialization.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @danielsilverstone-ct]](https://gitlab.com/danielsilverstone-ct) on Apr 16, 2019, 11:39

Right, so long as it's useful/needed at the scheduler point, we can calculate it then.  The main issue here is that things which do not need it *at all* are blocked on it (e.g. `bst workspace list`)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 16, 2019, 11:52

> Right, so long as it's useful/needed at the scheduler point, we can calculate it then. The main issue here is that things which do not need it at all are blocked on it (e.g. `bst workspace list`)

Rationally, I really think the only place it is worth conditionalizing is when we force *recalculate* the cache size at scheduler startup - this would avoid adding many conditionals where the gain in speed is negligible.

How much does it cost for `bst workspace list` to either of the following ?
* Read the cache_size file
* Perform `du` operation on an empty directory

Also, from recollection, the cache size resolving (as mentioned above) is usually only ever lazily resolved before scheduler starts, by way of cas.get_cache_size(), and that was only ever called on demand (i.e. if we ever want to print the cache size in advance of running the scheduler). So even the above operations normally should never happen unless asked for for some purpose.

Now some churn has happened and we have a `Cas` and a `CacheQuota` object belonging to the `Context`, maybe the mistake we made was to force resolve all this stuff and give it to `Context` which is needed for every command invocation.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Apr 24, 2019, 09:44

[[Gitlab user @danielsilverstone]](https://gitlab.com/danielsilverstone)\-ct [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) 

I investigated more.

When we ran the tests I didn't have a cache file for whatever reason (and it never got created/updated when doing successive 'shows').

Cleaning the cache entirely and retrying, I now have the file, so the problem is temporarily fixed. If I can find a way of reproducing I'll open a new issue for that. Laoding time is not under one second.

I still do think that we should keep this issue and load as many thing as lazily as possible. Does that make sense?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 24, 2019, 09:56

> I still do think that we should keep this issue and load as many thing as lazily as possible. Does that make sense?

Certainly !

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jennis]](https://gitlab.com/jennis) on Apr 24, 2019, 11:18

mentioned in merge request !1311

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Nov 5, 2019, 15:50

This is not relevant anymore. Casd is doing this out of band, and we don't suffer from it on the BuildStream side

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Nov 5, 2019, 15:50

closed
